pkg/configstore/store.go

```diff
@@ -0,0 +1,206 @@
+// trigger eager loading as per default functionality.
+//
+// There are two ways to get config data, via `Get` or `Load` methods. `Get`
+// will parse the config value and return it as `interface{}` type, while `Load`
+// will accept destination pointer and use `mapstructure.Decode` to destructure
+// the value.
+//
+// TODO: since `Store` is meant to be a long-lived object we need to keep track
+//       of updated configs and a possibility to manually reset the cache.
+type Store struct {
+	Data  *sync.Map
+	Dir   string
+	Hints map[string]string
+	sync.Mutex
+
+	ListFiles  func(string) ([]string, error)
+	ParseFile  func(string, interface{}) error
+	FileExists func(string) (bool, error)
+}
+
+func listFiles(dirname string) ([]string, error) {
+	files, err := ioutil.ReadDir(dirname)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to list directory %v: %v", dirname, err)
+	}
+
+	ret := make([]string, len(files))
+	for idx, file := range files {
+		ret[idx] = file.Name()
+	}
+
+	return ret, nil
+}
+
+func parseFile(path string, value interface{}) error {
+	reader, err := os.Open(path)
+	defer reader.Close()
+	if err != nil {
+		return fmt.Errorf("Failed to open %s: %v", path, err)
+	}
+
+	return json.NewDecoder(reader).Decode(value)
+}
+
+func fileExists(path string) (bool, error) {
+	_, err := os.Stat(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+// NewStore creates a new config store
+// `dir` is a directory from where data will be loaded
+// `hints` is a dictionary to find which file to load for
+// a given key
+func NewStore(dir string, hints map[string]string) *Store {
+	if hints == nil {
+		hints = map[string]string{}
+	}
+	return &Store{
+		Data:       &sync.Map{},
+		Dir:        dir,
+		Hints:      hints,
+		ListFiles:  listFiles,
+		ParseFile:  parseFile,
+		FileExists: fileExists,
+	}
+}
+
+// Decode `path` contents using `json`, lock the store mutex, merge loaded data
+// into s.Data, unlock the mutex
+func (s *Store) loadPath(path string) error {
+	var value map[string]interface{}
+	err := s.ParseFile(path, &value)
+	if err != nil {
+		return fmt.Errorf("Failed to parse %s: %v", path, err)
+	}
+
+	s.Lock()
+	defer s.Unlock()
+	for key, val := range value {
+		s.Data.Store(key, val)
+	}
+
+	return nil
+}
+
+// Walk `s.dir` and `s.loadPath` all the files
+func (s *Store) loadAll() error {
+	files, err := s.ListFiles(s.Dir)
+	if err != nil {
+		return fmt.Errorf("Failed to list %s: %v", s.Dir, err)
+	}
+
+	for _, file := range files {
+		filepath := path.Join(s.Dir, file)
+		err := s.loadPath(filepath)
+		if err != nil {
+			return fmt.Errorf("Failed to load %s: %v", filepath, err)
+		}
+	}
+	return nil
+}
+
+var extensions = []string{"json", "yaml"}
+
+// Look for `file`.json or `file`.yaml, if not found try loading all files and
+// print a warning about hints
+func (s *Store) load(file string, fromHint bool) error {
```

Comment:
is there a chance that we'll have a key that matches a filename, but that file won't contain the key?
it might be worth checking if `s.Data` contains the key after we load the file, and fallback to loading them all if not

Reply:
good spot, that was my intention but i guess as i moved things around i broke it, will add a test to cover this